### PR TITLE
[veejoy] Add new extractor

### PR DESCRIPTION
**Pull request by thilobillerbeck**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Read [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site)
- [x] Read [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) and adjusted the code to meet them
- [x] Covered the code with tests (note that PRs without tests will be REJECTED)
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This pull request adds support to download videos from https://www.veejoy.de/, a german video site operated and based around the company behind Europa Park. The extractor can be used to download either movies or episodes of series. Since the site seems to be built around nextjs, the extractor searches for `__NEXT_DATA__` and works its way towards the medias metadata and location from there.

Currently the most basic features are supported and the installer defaults to the uncompressed original footage, which surprisingly is lying at a public endpoint in their CDN. A method to extract the compressed chunk-based and compressed video is already in place (`get_video_formats`), but since the CDN seems to use some kind of cookie based checksum verification while downloading, it forbids me to download the chunks via ffmpeg. Unfortunately, I couldn't get my head around the problem yet, but as said, the extractor works with the uncompressed version.

### CODE REVIEW

**Comment by dirkf** [comment]
Please **fully complete the template**. Read all the bold text instructions and set the check boxes accordingly.

**Comment by thilobillerbeck** [comment]
Well, seems like me being stupid at copying the text I prefilled in another editor. Now it should be complete. 👍

**Comment by dirkf** [comment]
Finally, as there's no supporting issue, please fill in the [Site support issue template](https://github.com/ytdl-org/youtube-dl/issues/new?assignees=&labels=site-support-request&template=2_site_support_request.md) and paste the text into this thread.

**Review by dirkf — CHANGES_REQUESTED** [review]
Thanks for your work!

I've made some suggestions. Generally, `"string"` -> `'string'`. Maybe the m3u8 downloads can be revived.

**Comment by thilobillerbeck** [comment]
Thanks for replying to the PR so fast. :) I applied your suggestions and cleaned up some other stuff accordingly. :+1:

**Comment by thilobillerbeck** [comment]
> Finally, as there's no supporting issue, please fill in the [Site support issue template](https://github.com/ytdl-org/youtube-dl/issues/new?assignees=&labels=site-support-request&template=2_site_support_request.md) and paste the text into this thread.




## Checklist



- [x] I'm reporting a new site support request
- [x] I've verified that I'm running youtube-dl version **2021.12.17**
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that none of provided URLs violate any copyrights
- [x] I've searched the bugtracker for similar site support requests including closed ones


## Example URLs



- Single: https://www.veejoy.de/en/series/o-the-construction-documentary/how-it-all-started
- Single: https://www.veejoy.de/en/movies/making-of-the-secret

## Description



See original post.

File: youtube_dl/extractor/veejoy.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+            'uploader': 'MACK Media'
+        }
+    }]
+
+    def get_video_id(self, url):
+        return self._match_id(url)
+
+    def get_video_data(self, url, video_id):
+        webpage = self._download_webpage(url, video_id)
+        next_data = self._html_search_regex(r'<script id="__NEXT_DATA__" type="application/json">([^<]+)</script>', webpage, 'next_data')
```
[the remark this answers is not in the document]
Comment (thilobillerbeck) [reply youtube_dl/extractor/veejoy.py]: Applied suggestion

File: youtube_dl/extractor/veejoy.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        if thumb_3_4:
+            thumbnails.append({
+                'url': thumb_3_4["srcSet"][1].split(" ")[0],
+            })
+
+        thumb_16_9 = video_data["teaserImage"]["16_9"]
+        if thumb_16_9:
+            thumbnails.append({
+                'url': thumb_16_9["srcSet"][1].split(" ")[0],
+            })
```
[the remark this answers is not in the document]
Comment (thilobillerbeck) [reply youtube_dl/extractor/veejoy.py]: Applied suggestion

File: youtube_dl/extractor/veejoy.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        'info_dict': {
+            'id': 'o-the-construction-documentary/the-building-site-grows',
+            'ext': 'mp4',
+            'title': 'Bau-„Leiter“',
+            'description': 'Auf der Baustelle ist viel passiert. Patrick und Lukas bekommen ein Update vom Bauleiter, erklären technische Grundlagen am „lebenden Objekt“ und stellen sich einer Onride-Challenge.',
+            'uploader': 'MACK Media'
+        }
+    }]
+
+    def get_video_id(self, url):
```
[the remark this answers is not in the document]
Comment (thilobillerbeck) [reply youtube_dl/extractor/veejoy.py]: Applied suggestion

File: youtube_dl/extractor/veejoy.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    def get_video_data(self, url, video_id):
+        webpage = self._download_webpage(url, video_id)
+        next_data = self._html_search_regex(r'<script id="__NEXT_DATA__" type="application/json">([^<]+)</script>', webpage, 'next_data')
+        return json.loads(next_data)["props"]["pageProps"]["media"]
+
+    def get_producer(self, video_data):
+        if "item" in video_data["studioDetails"]:
+            return video_data["studioDetails"]["item"]["title"]
+        else:
+            return "Veejoy"
```
[the remark this answers is not in the document]
Comment (thilobillerbeck) [reply youtube_dl/extractor/veejoy.py]: Applied suggestion

File: youtube_dl/extractor/veejoy.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+                    'width': asset_format["transcodingFormat"]["videoWidth"],
+                    'quality': asset_format["transcodingFormat"]["id"],
+                    'language': asset_format["language"],
+                })
+
+        return formats
+
+    def _real_extract(self, url):
+        video_id = self.get_video_id(url)
+        video_data = self.get_video_data(url, video_id)
```
[the remark this answers is not in the document]
Comment (thilobillerbeck) [reply youtube_dl/extractor/veejoy.py]: Applied suggestion

File: youtube_dl/extractor/veejoy.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        producer = self.get_producer(video_data)
+        thumbnails = self.get_thumbnails(video_data)
+        final_asset = self.get_original_file_url(video_data, video_id)
```
[the remark this answers is not in the document]
Comment (thilobillerbeck) [reply youtube_dl/extractor/veejoy.py]: Applied suggestion

File: youtube_dl/extractor/veejoy.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+        return {
+            'url': final_asset.get("contentUrl"),
+            'id': video_id,
+            'title': video_data.get("title"),
+            'description': video_data.get("shortDescription"),
+            'duration': video_data.get("mediaDuration"),
```
[the remark this answers is not in the document]
Comment (thilobillerbeck) [reply youtube_dl/extractor/veejoy.py]: Added apropriate handlers to each field

File: youtube_dl/extractor/veejoy.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+    def _real_extract(self, url):
+        video_id = self.get_video_id(url)
+        video_data = self.get_video_data(url, video_id)
+        producer = self.get_producer(video_data)
+        thumbnails = self.get_thumbnails(video_data)
+        final_asset = self.get_original_file_url(video_data, video_id)
+
+        return {
+            'url': final_asset.get("contentUrl"),
```
[the remark this answers is not in the document]
Comment (thilobillerbeck) [reply youtube_dl/extractor/veejoy.py]: Applied suggestion

File: youtube_dl/extractor/veejoy.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+        return thumbnails
+
+    def get_asset_ref(self, video_data):
+        for mediaAsset in video_data["mediaAssets"]:
+            if mediaAsset["type"] == "SOURCE":
+                return mediaAsset["assetReference"]
+
+    def get_asset_formats(self, video_data, video_id):
+        asset = self._download_json("https://www.veejoy.de/api/service/get-media-summary?mediaIri=" + self.get_asset_ref(video_data) + "&locale=en", video_id)
```
[the remark this answers is not in the document]
Comment (thilobillerbeck) [reply youtube_dl/extractor/veejoy.py]: Applied suggestion

File: youtube_dl/extractor/veejoy.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+        return asset["assetFormats"]
+
+    def get_original_file_url(self, video_data, video_id):
+        for asset_format in self.get_asset_formats(video_data, video_id):
+            if asset_format["mimeType"] == "video/mp4":
```
[the remark this answers is not in the document]
Comment (thilobillerbeck) [reply youtube_dl/extractor/veejoy.py]: Applied suggestion

File: youtube_dl/extractor/veejoy.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+        if thumb_16_9:
+            thumbnails.append({
+                'url': thumb_16_9["srcSet"][1].split(" ")[0],
+            })
+
+        return thumbnails
+
+    def get_asset_ref(self, video_data):
+        for mediaAsset in video_data["mediaAssets"]:
+            if mediaAsset["type"] == "SOURCE":
```
[the remark this answers is not in the document]
Comment (thilobillerbeck) [reply youtube_dl/extractor/veejoy.py]: Applied suggestion

File: youtube_dl/extractor/extractors.py [hunk 12]
```diff
@@ -1661,3 +1661,4 @@
 )
 from .zoom import ZoomIE
 from .zype import ZypeIE
+from .veejoy import VeejoyIE
```
[the remark this answers is not in the document]
Comment (thilobillerbeck) [reply youtube_dl/extractor/extractors.py]: Applied suggestion

File: youtube_dl/extractor/veejoy.py [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+        formats = []
+
+        for asset_format in asset_formats:
+            if asset_format["mimeType"] == "application/vnd.apple.mpegurl":
+                formats.append({
+                    'url': asset_format["contentUrl"],
+                    'width': asset_format["transcodingFormat"]["videoWidth"],
+                    'quality': asset_format["transcodingFormat"]["id"],
+                    'language': asset_format["language"],
+                })
```
[the remark this answers is not in the document]
Comment (thilobillerbeck) [reply youtube_dl/extractor/veejoy.py]: Applied suggestions

**Comment by thilobillerbeck** [comment]
Hi, I finally managed to go through all your suggestions and chek against what I've applied.

File: youtube_dl/extractor/veejoy.py [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+        q = qualities(['hq', 'mq', 'lq'])
+
+        for asset_format in asset_formats:
+            f_url = url_or_none(asset_format.get('contentUrl'))
+            if not f_url:
+                continue
+            ext = determine_ext(f_url)
+            transcodingFormat = try_get(asset_format, lambda x: x['transcodingFormat'], dict) or {}
+
+            if transcodingFormat == {}:
```
Comment (dirkf) [inline youtube_dl/extractor/veejoy.py:116]:
```suggestion
            transcodingFormat = try_get(asset_format, lambda x: x['transcodingFormat'], dict)

            if not transcodingFormat:
```

File: youtube_dl/extractor/veejoy.py [hunk 15]
```diff
@@ -0,0 +1,160 @@
+# coding: utf-8
+from __future__ import unicode_literals
+
+from .common import InfoExtractor
+from ..compat import compat_str
+from ..utils import (
+    ExtractorError,
```
Comment (dirkf) [inline youtube_dl/extractor/veejoy.py:7]:
```suggestion
```

**Pull request closed without merging**